Add `torch.unflatten`

k223kim opened the pull request:
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #288. FYI I used `view` rather than `reshape`.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

🚀🚀🚀

nikitaved commented on thunder/tests/opinfos.py:
```diff
@@ -3609,6 +3609,47 @@ def reshape_sample_generator(op, device, dtype, requires_grad, **kwargs):
+        ((4, 2), 0, (2, 2)),
+        ((125, 3), 0, (25, 5)),
+        ((25, 25), -1, (1, 5, 5)),
+        ((16, 32), -1, (2, 4, 1, 4)),
+        ((4, 5, 6), -1, (6, 1, 1, 1)),
```
PyTorch allows `-1` sizes. Let's test these as well?

mruberry replied:
Adding tests for partial shape specification is great. Let's also add some tests of unflattening dimensions which are not the outermost or innermost, like `(4, 12, 6), 1, ...`

k223kim replied:
Sure thing! I wasn't aware that `-1` sizes were allowed. Thanks so much!

mruberry commented on thunder/torch/__init__.py:
```diff
@@ -782,6 +782,17 @@ def reshape(a: TensorLike, /, *shape: int) -> TensorLike:
 
 
+@torchsymbol(torch.unflatten, is_method=True)
+def unflatten(a: TensorLike, /, dim: int, sizes=tuple[int, ...]) -> TensorLike:
```
This signature looks pretty nice, but `sizes-tuple[int, ...]` is assigning the type annotation, and the type annotation is overly strict because any `Sequence` is OK, so it should be:

```
sizes: Sequence[int]
```

k223kim replied:
You are right! Didn't know PyTorch supported lists and other sequence as well. Will update :) Thanks!

mruberry commented on thunder/torch/__init__.py:
```diff
@@ -782,6 +782,17 @@ def reshape(a: TensorLike, /, *shape: int) -> TensorLike:
 
+@torchsymbol(torch.unflatten, is_method=True)
+def unflatten(a: TensorLike, /, dim: int, sizes=tuple[int, ...]) -> TensorLike:
+    utils.check(
```
Is this error check necessary, or will the error be caught by `view`?

k223kim replied:
This error will be caught by `view`, specifically, in `reshape`! However, it throws an error that seems specific for `reshape`. For instance, in the case of `SampleInput(input_tensor, 0, ())`, the error message will be something like the following:
```
Attempting to reshape a.shape=(4, 4) to shape=(4,), but a.numel=16 is different from the number of elements in shape, 4
```
I thought this error message is clearer? What do you think? If not, I can definitely remove this.

mruberry commented on thunder/torch/__init__.py:
```diff
@@ -782,6 +782,17 @@ def reshape(a: TensorLike, /, *shape: int) -> TensorLike:
+        lambda: f"unflatten() sizes must be non-empty",
+        RuntimeError,
+    )
+    dim = utils.canonicalize_dim(a.ndim, dim)
```
Nice canonicalization

mruberry commented on thunder/tests/opinfos.py:
```diff
@@ -3609,6 +3609,47 @@ def reshape_sample_generator(op, device, dtype, requires_grad, **kwargs):
+    make = partial(make_tensor, dtype=dtype, device=device)
+    input_tensor = make(4, 4)
+    dim = 3
+    yield (SampleInput(input_tensor, 0, ()), RuntimeError, r"unflatten\(\) sizes must be non-empty")
```
Two more error cases that are interesting:

- the unflattened shape is fully specified and invalid. For example, a dimension of length 12 and a request to unflatten it to `(4, 4)`
- the unflattened shape is partially specified and invalid. For example, a dimension of length 12 and a request to unflatten it to `(-1, 5)`.

I imagine that the call to `view()` will handle these errors properly for these cases, but it's still nice to check them from `unflatten()`

k223kim replied:
Awesome idea! I thought error cases should only include the ones caught by this specific function (unflatten) but it makes sense to check these.

mruberry commented on thunder/tests/opinfos.py:
```diff
@@ -3609,6 +3609,47 @@ def reshape_sample_generator(op, device, dtype, requires_grad, **kwargs):
+def unflatten_sample_generator(op, device, dtype, requires_grad, **kwargs):
+    make = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    cases = (
```
A short comment like 

```
# shape, dim, unflatten_shape
```

before `cases` would improve readability

mruberry commented on thunder/tests/opinfos.py:
```diff
@@ -3609,6 +3609,47 @@ def reshape_sample_generator(op, device, dtype, requires_grad, **kwargs):
+    cases = (
+        ((4,), 0, (4,)),  # no-op
+        ((2, 2), -1, (2,)),  # no-op
+        ((1, 2, 1), 1, (2,)),  # no-op
```
Nice no-op testing

mruberry reviewed (commented):
Looking good so far, @k223kim! I made some comments for your review

k223kim commented on thunder/torch/__init__.py:
```diff
@@ -782,6 +782,17 @@ def reshape(a: TensorLike, /, *shape: int) -> TensorLike:
+        RuntimeError,
+    )
+    dim = utils.canonicalize_dim(a.ndim, dim)
+    return a.view(tuple(a.shape[:dim]) + tuple(sizes) + tuple(a.shape[dim + 1 :]))
```
Hmm, if I am understanding this correctly, when doing `a.shape`, it returns `torch.Size([...])` which is a `torch.Size` class. Won't we want an explicit conversion to `tuple`?

k223kim replied:
Oh I see. You are right. I completely missed that. Will update this right now :) Thank you!

k223kim commented:
Hey @mruberry ! Thanks for the detailed feedback as always. Updated some lines and I have added more test cases as per the comments. I do have some questions regarding some of your comments and would appreciate it if you can take a look!

mruberry reviewed (approved):
<img src="https://media2.giphy.com/media/3WCNY2RhcmnwGbKbCi/giphy.gif"/>

Awesome! Nice reuse of `view`

fyi @Borda and @t-vi to merge
